woocommerce/woocommerce-ios
# Product Selector: Add single selection mode and simple selection handling mode

Closes: #11585


## Description


This PR updates the existing product selector view (notably available via Create new order > Add Products) with two optional configurations:

- **SelectionHandling:** How the selector should behave if a product is selected. Either `.simple` or `.normal`.
  - `.simple`: All type of products are immediately selected
  - `.normal`: variable or bundle products are not immediately selected, but have extra action (opening another screen). This is the default behavior.
- **SelectionMode:** How many product(s) can be selected in one session. Either `.single` or `.multiple`.
  - `.single`: Selector only allows selecting a product 
  - `.multiple`: Selector allows selecting more than one products. This is the default behavior.
   
This matrix explains how these configurations work in combination with each other:

<table border="1">
  <tr>
    <td></td>
    <td><b>SelectionMode: single</b></td>
    <td><b>SelectionMode: multiple</b></td>
  </tr>
  <tr>
    <td><b>SelectionHandling: simple</b></td>
    <td>This is the requirement for Blaze i3: <ul><li>Only allow selecting one product at a time.</li> <li>Selecting a variable product immediately selects it, no option to select variations.</li> <li>Selecting a bundle product immediately selects it, no option to change configuration.</li></td>
    <td>Allows selecting multiple products in a simple (quick) way, ignoring variations or configurations. Not yet used anywhere but is now possible.</td>
  </tr>
  <tr>
    <td><b>SelectionHandling: normal</b></td>
    <td>Only allows selecting one product at a time, but default variable and bundle product selection behavior are respected. Not yet used anywhere but is now possible.</td>
    <td>This is the product selector's existing default behavior.<ul><li>Allows selecting multiple products.</li> <li>Selecting a variable product opens variations selector.</li> <li>Selecting a bundle product opens its configuration screen.</li></td>
  </tr>
</table>

As mentioned above, technically only the simple handling+single selection mode case is currently needed for the Blaze i3 project, but support for all combinations is also made available here. This is also to match the Android counterpart in  https://github.com/woocommerce/woocommerce-android/pull/10490

## Not included in this PR
The Blaze i3 design also requires us to update the selector title, and change the back button to be an arrow. This as far as I can tell is already changeable.

## Testing instructions


Use this patch:

```
diff --git a/WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift b/WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift
index 2959b3eddb..1b0627dea1 100644
--- a/WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift	
+++ b/WooCommerce/Classes/ViewRelated/Orders/Order Creation/EditableOrderViewModel.swift	
@@ -1710,6 +1710,8 @@ private extension EditableOrderViewModel {
                     siteID: siteID,
                     selectedItemIDs: selectedProductsAndVariationsIDs,
                     purchasableItemsOnly: true,
+                    selectionHandlingMode: .simple,
+                    selectionMode: .single,
                     storageManager: storageManager,
                     stores: stores,
                     toggleAllVariationsOnSelection: false,
```

This patch sets the product selector in Add Order screen to behave how it should be in Blaze i3.

### 1. Blaze-related testing:

1. Go to Order > Add Order > Add Product,
2. Ensure that the button at the bottom always says "Done",
3. Ensure that the "Clear selection" option is not shown on top left,
4. Select a simple product,
5. Select another simple product, ensure that the previous product is deselected.
6. Select a variable product, ensure that the product is immediately selected, and does not navigate to variation selector. Ensure that the previous product is deselected.

#### extra testing for product bundle:
7. Install the extension https://woo.com/products/product-bundles/
8. Create a Product bundle,
9. Go to Order > Add Order > Add Product,
10. Select a simple product,
11. Select the bundle product, ensure that the product is immediately selected and does not navigate to configuration screen. Ensure that the previous product is deselected.

Here is a video demonstrating the testing above:

https://github.com/woocommerce/woocommerce-ios/assets/266376/c11d3f65-ab78-4e5b-b160-994a17a72aa3



### 2. Single selection mode with normal selection handling
1. Go to the patched code above, change `selectionHandlingMode: .simple,` to `selectionHandlingMode: .normal,`
2. Go to Order > Add Order > Add Product,
3. Select a simple product and ensure it is selected,
4. Select a variable product, ensure that it opens to variation product selector,
5. On the variation product selector screen, ensure that the "Clear selection" option is not shown on top left,
6. Try selecting some variation products, ensure that only one product can be selected,
7. Go back,
8. Ensure that the variation product has the partially selected checkbox, and the simple product from step 3 is not selected anymore.
9. Extra: select a bundle product and ensure its configuration screen is shown instead of it being selected immediately.

Here is a video demonstrating the testing above:

https://github.com/woocommerce/woocommerce-ios/assets/266376/603d4c25-0227-4546-832d-d70f06ae922a



### known issue for this case related to bundle product.
After selecting a simple/variation product, then selecting a bundle product, then the simple/variation product will remain selected. Thus, it breaks the single selection mode rule.

This is because when the bundle product is selected (by saving on its configuration screen), there is no callback that can be used by `ProductSelectorViewModel` to update its selection state. Instead the saving callback is provided by the parent view of Product Selector (e.g: `EditableOrderViewModel`) to the `ConfigurableBundleProductViewModel` ([here for example](https://github.com/woocommerce/woocommerce-ios/blob/0ed961fc79be148175135bac06a591665845b1e3/WooCommerce/Classes/ViewRelated/Orders/Order%20Creation/EditableOrderViewModel.swift#L688-L690)).

This is an edge case and not used anywhere in the app yet, so I think can be handled as a separate issue if needed.

---
